### PR TITLE
`stable-symbolic`: Properly initialize immutable `malloc`'d memory

### PR DESCRIPTION
**Pull request by RyanGlScott**
Previously, `stable-symbolic` would only initialize mutable memory with symbolic bytes when calling `malloc`. `malloc` can also be used to initialize immutable memory, however, so we must also account for this possibility when `stable-symbolic` is enabled.

To this end, we create a new `doConstStoreStableSymbolic` function that is like `doStoreStableSymbolic`, but does not check if the memory being written to is mutable. (This is very much in the spirit of the existing `doArrayStore`/`doArrayConstStore`, `storeRaw`/`storeConstRaw`, and `mallocRaw`/`mallocConstRaw` split.) We then pick between `doStoreStableSymbolic` and `doConstStoreStableSymbolic` in the override for `malloc` depending on whether the memory is mutable or immutable, respectively.

See GaloisInc/saw-script#1691 for the motivation behind this bugfix.

### CODE REVIEW

File: crucible-llvm/src/Lang/Crucible/LLVM/MemModel.hs [hunk 1]
```diff
@@ -1070,6 +1101,30 @@ doStoreStableSymbolic bak mem ptr mbSz alignment = do
...
     Nothing -> doArrayStoreUnbounded bak mem ptr alignment bytes
 
+-- | Store a fresh symbolic value of the appropriate size in the supplied
+-- pointer. This is used in various spots whenever 'laxLoadsAndStores' is
+-- enabled and 'indeterminateLoadBehavior' is set to 'StableSymbolic'.
+--
+-- Precondition: the pointer is valid and points to a mutable or immutable
+-- memory region. Therefore it can be used to initialize read-only memory
+-- regions.
+doConstStoreStableSymbolic ::
```
[the remark this answers is not in the document]
Comment (robdockins) [reply crucible-llvm/src/Lang/Crucible/LLVM/MemModel.hs:1111]: I think having separate functions is deliberate. The `Const` variants are only ever intended to be used in initialization, and the intention is to make it hard to accidentally chose the one that is abstraction-breaking (can write to immutable memory).

File: crucible-llvm/src/Lang/Crucible/LLVM/MemModel.hs [hunk 2]
```diff
@@ -710,10 +711,11 @@ doMallocSize sz bak allocType mut loc mem alignment = do
   let ptr   = LLVMPointer blk z
   let mem'  = mem{ memImplHeap = heap' }
   mem'' <- if laxLoadsAndStores ?memOpts
-                && mut == G.Mutable
                 && allocType == G.HeapAlloc
                 && indeterminateLoadBehavior ?memOpts == StableSymbolic
-           then doStoreStableSymbolic bak mem' ptr sz alignment
+           then if mut == G.Mutable
```
[the remark this answers is not in the document]
Comment (RyanGlScott) [reply crucible-llvm/src/Lang/Crucible/LLVM/MemModel.hs]: A fair point. `doStoreStableSymbolic` would generate an additional assertion that the memory is mutable, but since we know whether the allocated memory is mutable or not from `doMalloc`'s argument, there's not much point in generating the assertion in the first place.

There are also two other call sites to `doStoreStableSymbolic` that warrant further scrutiny: one in `doAlloca` and one in `doInvalidate`. In `doAlloca`, we know that we are always allocating mutable memory, so I think `doConstStoreStableSymbolic` makes more sense there. In `doInvalidate`, we don't necessary know at a glance if the memory being invalidated is mutable or not, so I think it makes more sense to call `doStoreStableSymbolic` to generate the check-if-memory-is-mutable assertion.

In short, I've pushed a change that changes the uses of `doStoreStableSymbolic` in `do{Malloc,Alloca}` to `doConstStoreStableSymbolic` while leaving `doInvalidate` alone.

**Review by robdockins — APPROVED** [review]
